File: javascript/smartlist/smartlist-api-mongodb-project/src/modules/question/question-strategy/mcq-multiple-strategy.ts
import { IQuestionStrategy } from '../question-interface/question-interface';
import { QuestionReqDto } from '../dto/request/question.req.dto';
import { QuestionResponseCodes } from '../question.response.codes';

export class McqMultipleStrategy implements IQuestionStrategy {
  constructor(protected questionReqDto: QuestionReqDto) {}
  validate() {
    if (this.questionReqDto.answer_options.length < 2) {
      throw QuestionResponseCodes.INVALID_ANSWER_OPTIONS_FOR_QUESTION_TYPE;
    }

    let correctCount = 0;
    this.questionReqDto.answer_options.forEach((ao) => {
      correctCount += ao.correct ? 1 : 0;
    });
    if (correctCount < 2) {
      throw QuestionResponseCodes.INVALID_ANSWER_OPTION;
    }
  }
}

File: java/smartlist/src/main/java/com/smartlist/module/question/questionstrategy/McqMultipleStrategy.java
package com.smartlist.module.question.questionstrategy;

import com.smartlist.module.question.QuestionResponseCodes;
import com.smartlist.module.question.dto.request.QuestionReqDTO;
import com.smartlist.module.question.interfaces.IQuestionStrategy;

public class McqMultipleStrategy implements IQuestionStrategy {

    private final QuestionReqDTO questionReqDto;

    public McqMultipleStrategy(final QuestionReqDTO questionReqDto){
        this.questionReqDto = questionReqDto;
    }

    @Override
    public void validate() {
        if (this.questionReqDto.getAnswerOptions().size() < 2) {
            throw QuestionResponseCodes.INVALID_ANSWER_OPTIONS_FOR_QUESTION_TYPE;
        }

        int correctCount = this.questionReqDto.getAnswerOptions().stream().mapToInt(ao -> Boolean.TRUE.equals(ao.getCorrect()) ? 1 : 0).sum();
        if (correctCount < 2) {
            throw QuestionResponseCodes.INVALID_ANSWER_OPTION;
        }
    }
}
